### PR TITLE
Feature/user invitation

### PR DESCRIPTION
**Pull request by apeksha2697**
created this PR to add a feature to send user invitation.
changes made:
1. created edit user modal
2. added invitation_status column to users table
3. created modal to invite new user
4. updated invitation controller methods
5. created a list of users pending invitation 
6. resend invitation mail to user
7. tests for invitation controller 
8. added pagy gem for pagination

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
+  include Pagy::Backend
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
```
Comment (BandanaPandey) [inline app/controllers/application_controller.rb:4]: What this much line space between befor_action and action ?

File: app/controllers/users/invitations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Users::InvitationsController < Devise::InvitationsController
+  before_action :authenticate_inviter!, only: [:new, :create, :pending]
+  
+  def create
+    @user = User.new(user_params)
```
Comment (BandanaPandey) [inline app/controllers/users/invitations_controller.rb:5]: refactor, using user_params map functionality

File: app/controllers/users/invitations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    @user.organisation_id = current_user.organisation_id
+    @user.invitation_status = "Pending"
+    @user.password = "password"
+      if @user.valid? && @user.invite!
```
Comment (BandanaPandey) [inline app/controllers/users/invitations_controller.rb]: @notice = "Invitation Sent" if @user.valid? && @user.invite!
render "new"

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -2,23 +2,37 @@ class UsersController < ApplicationController
...
-    @users = User.find(params[:id])
-    @users.destroy
+    @user = User.find(params[:id])
+    @user.destroy
    
     redirect_to users_path
   end
 
+  def search
+    @pagy, @users = pagy(User.where("first_name Like ? OR last_name Like ?", params[:name].titlecase, params[:name].titlecase))
```
Comment (BandanaPandey) [inline app/controllers/users_controller.rb]: refactor into two lines

Comment (BandanaPandey) [reply app/controllers/users_controller.rb]: use common variable for params[:name].titlecase

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -2,23 +2,37 @@ class UsersController < ApplicationController
...
+  end
+
   private
   
   def user_params
-    params.require(:user).permit(:first_name, :last_name, :gender, :role_id, :organisation_id, :email, :password, :password_confirmation)
```
Comment (BandanaPandey) [inline app/controllers/users_controller.rb:22]: params.require(:user)
             .permit(:first_name, :last_name, :gender, :age, :role_id, :organisation_id, 
             :email, :password,:password_confirmation)

File: app/views/users/_edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        <%= f.number_field :age, placeholder: "Age", autofocus: true %><br />
+    </div>
+    <div class="modal-footer">
+      <%= f.submit "Update" %> 
+      <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>    
+    </div>
+    </div>
+      <% end %>
+  </div>
+</div>
```
Comment (BandanaPandey) [inline app/views/users/_edit.html.erb]: new line solve

File: app/views/users/invitations/pending.html.erb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  <table>
+    <tr>
+      <th>Name</th>
+      <th>Email</th>
+      <th>Role</th>
+      <th>Status</th>
+      <th></th>
+      <th></th>
+    </tr>
+    <% @user.all.each do |user| %>
```
Comment (BandanaPandey) [inline app/views/users/invitations/pending.html.erb]: rails view collection and partial use

File: app/views/users/index.html.erb [hunk 8]
```diff
@@ -1,47 +1,38 @@
...
   <table>
     <tr>
       <th>Name</th>
       <th>Email</th>
       <th>Role</th>
+      <th>Status</th>
       <th></th>
       <th></th>
     </tr>
     <% @users.all.each do |user| %>
```
Comment (BandanaPandey) [inline app/views/users/index.html.erb]: @users.each

File: app/views/users/index.html.erb [hunk 9]
```diff
@@ -1,47 +1,38 @@
...
-      <% if user.organisation_id == current_user.organisation_id %>
+      <% if user.organisation_id == current_user.organisation_id && (user.invitation_status == "Accepted" || user.invitation_status == nil) %>
```
Comment (BandanaPandey) [inline app/views/users/index.html.erb]: refactor

File: app/views/users/index.html.erb [hunk 10]
```diff
@@ -1,47 +1,38 @@
...
           <td><%= user.email %></td>
           <td><%= user.role.name %></td>
-          <td><%= link_to sanitize('<i class="fas fa-pencil-alt" aria-hidden="true"></i>'), edit_user_path(user)  %></td>
-          <td><%= button_to user_path(user), method: :delete, data: { confirm: 'Are you sure?' } do %>
-            <i class="fa fa-trash" aria-hidden="true"></i>
-            <% end %></td>                   
+          <td><%= user.invitation_status %></td>
+          <td><%= link_to edit_user_path(user.id), remote: true  do %>
+          <span class="fas fa-pencil-alt" aria-hidden="true"></span>
+          <%end%></td>
```
Comment (BandanaPandey) [inline app/views/users/index.html.erb]: fix it

File: app/views/users/invitations/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    <%= f.email_field :email, readonly: true %><br/>
+    <%= f.hidden_field :invitation_status, value: "Accepted" %>
+    <div class="error" ><%= resource.errors[:password][0] %></div>
+    <%= f.password_field :password, placeholder: "Password", autocomplete: "new-password"  %>
+    <div class="error" ><%= resource.errors[:password_confirmation][0] %></div>
+    <%= f.password_field :password_confirmation,placeholder: "Password Confirmation", autocomplete: "new-password" %>
+    <%= f.submit "Create Account" %>
+    <% end %>
+  </div>
+  </body>
```
Comment (BandanaPandey) [inline app/views/users/invitations/edit.html.erb]: align body tag

File: config/environments/test.rb [hunk 12]
```diff
@@ -40,6 +40,7 @@
   # The :test delivery method accumulates sent emails in the
   # ActionMailer::Base.deliveries array.
   config.action_mailer.delivery_method = :test
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (BandanaPandey) [inline config/environments/test.rb:43]: check this, or remove it

File: db/migrate/20201015091045_add_invitation_status_to_users.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class AddInvitationStatusToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :invitation_status, :string
+  end
+
```
Comment (BandanaPandey) [inline db/migrate/20201015091045_add_invitation_status_to_users.rb]: space ??

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (apeksha2697) [reply app/controllers/application_controller.rb:4]: removed extra space

File: app/controllers/users/invitations_controller.rb (hunk 2 above)
Comment (apeksha2697) [reply app/controllers/users/invitations_controller.rb:5]: done

File: app/controllers/users/invitations_controller.rb (hunk 3 above)
Comment (apeksha2697) [reply app/controllers/users/invitations_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (apeksha2697) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (apeksha2697) [reply app/controllers/users_controller.rb:22]: done

File: app/views/users/_edit.html.erb (hunk 6 above)
Comment (apeksha2697) [reply app/views/users/_edit.html.erb]: done

File: app/views/users/index.html.erb (hunk 8 above)
Comment (apeksha2697) [reply app/views/users/index.html.erb]: done

File: app/views/users/index.html.erb (hunk 9 above)
Comment (apeksha2697) [reply app/views/users/index.html.erb]: done

File: app/views/users/index.html.erb (hunk 10 above)
Comment (apeksha2697) [reply app/views/users/index.html.erb]: done

File: app/views/users/invitations/edit.html.erb (hunk 11 above)
Comment (apeksha2697) [reply app/views/users/invitations/edit.html.erb]: done

File: app/views/users/invitations/pending.html.erb (hunk 7 above)
Comment (apeksha2697) [reply app/views/users/invitations/pending.html.erb]: done

File: config/environments/test.rb (hunk 12 above)
Comment (apeksha2697) [reply config/environments/test.rb:43]: done

File: db/migrate/20201015091045_add_invitation_status_to_users.rb (hunk 13 above)
Comment (apeksha2697) [reply db/migrate/20201015091045_add_invitation_status_to_users.rb]: done

File: app/helpers/users_helper.rb [hunk 14]
```diff
@@ -11,4 +11,9 @@ def first_and_last_name_initials(user)
   def full_name(user)
     "#{user.first_name + " " + user.last_name}"
   end
+  
+  def all_users_in_organisation(user)
```
Comment (BandanaPandey) [inline app/helpers/users_helper.rb]: @apeksha2697  please think of proper method name, here you are saying all_users_in_organisation, but below you are checking on a single user

File: app/views/layouts/application.html.erb [hunk 15]
```diff
@@ -11,8 +11,8 @@
   </head>
 
   <body>
-  <%= render 'shared/navbar' %>
-  <%= render 'shared/sidepanel' %>
+    <%= render 'shared/navbar' %>
```
Comment (BandanaPandey) [inline app/views/layouts/application.html.erb:14]: again why there is tab on line 14, 15
Please remove tab space and use 2 line space

File: app/views/users/_all_users_list.html.erb [hunk 16]
```diff
@@ -0,0 +1,28 @@
+<table>
+  <tr>
+    <th>Name</th>
+    <th>Email</th>
+    <th>Role</th>
+    <th>Status</th>
+    <th></th>
+    <th></th>
+  </tr>
+  <% @users.each do |user| %>
```
Comment (BandanaPandey) [inline app/views/users/_all_users_list.html.erb]: @apeksha2697  Please use rails collection and partial on view,
read about this https://guides.rubyonrails.org/layouts_and_rendering.html#rendering-collections
and implement it in the similar way

File: app/views/users/_search_users_list.html.erb [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+  <tr>
+    <th>Name</th>
+    <th>Email</th>
+    <th>Role</th>
+    <th>Status</th>
+    <th></th>
+    <th></th>
+  </tr>
+  <% @users.each do |user| %>
+    <% if user.organisation_id == current_user.organisation_id %>
```
Comment (BandanaPandey) [inline app/views/users/_search_users_list.html.erb]: same move to different partial, use rails collection partial functionality

File: app/views/users/invitations/_new.html.erb [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+        <%= form_for(resource, as: resource_name, url: invitation_path(resource_name),remote: true, 
+                    html: { method: :post }) do |f| %>
+          <div id='user_invitation'></div>
+
+          <div class="form">
+            <div class="notice" >
+              <%= @notice %>
+            </div>
+            
+            <div class="error" ><%= resource.errors[:first_name][0] %></div>
```
Comment (BandanaPandey) [inline app/views/users/invitations/_new.html.erb]: closing div should be below the opening div, as well as properly indent it like what you have done above from line number 16 to line no 18

File: app/views/users/invitations/_new.html.erb [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+            <%= f.text_field :first_name, placeholder: "First Name"  %>
+            
+            <%= f.text_field :last_name, placeholder: "Last Name"  %>
+
+            <div class="error" ><%= resource.errors[:email][0] %></div>
```
Comment (BandanaPandey) [inline app/views/users/invitations/_new.html.erb]: same as above

File: app/views/users/invitations/_pending_users_list.html.erb [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+  <tr>
+    <th>Name</th>
+    <th>Email</th>
+    <th>Role</th>
+    <th>Status</th>
+    <th></th>
+    <th></th>
+  </tr>
+  <% @pending_users.each do |user| %>
+    <% if user.organisation_id == current_user.organisation_id %>
```
Comment (BandanaPandey) [inline app/views/users/invitations/_pending_users_list.html.erb]: same use different partial and collection rails view

File: config/environments/test.rb (hunk 12 above)
Comment (BandanaPandey) [reply config/environments/test.rb:43]: what have you done, You haven't removed this line, neither did you explain why it is there in test.rb ?

File: db/migrate/20201006121709_devise_invitable_add_to_users.rb [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+      t.index      :invitations_count
+      t.index      :invitation_token, unique: true # for invitable
+      t.index      :invited_by_id
+    end
+  end
+
+  def down
+    change_table :users do |t|
+      t.remove_references :invited_by, polymorphic: true
+      t.remove :invitations_count, :invitation_limit, :invitation_sent_at, :invitation_accepted_at, :invitation_token, :invitation_created_at
```
Comment (BandanaPandey) [inline db/migrate/20201006121709_devise_invitable_add_to_users.rb]: line is too big, break it into two lines

File: test/controllers/home_controller_test.rb [hunk 22]
```diff
@@ -2,7 +2,6 @@
 
 class HomeControllerTest < ActionController::TestCase
   def setup
-    super
     @user = create :user
   end
```
Comment (BandanaPandey) [inline test/controllers/home_controller_test.rb]: remove this file, it is of no use here

Comment (apeksha2697) [reply test/controllers/home_controller_test.rb]: done

File: db/migrate/20201006121709_devise_invitable_add_to_users.rb (hunk 21 above)
Comment (apeksha2697) [reply db/migrate/20201006121709_devise_invitable_add_to_users.rb]: done

File: app/helpers/users_helper.rb (hunk 14 above)
Comment (apeksha2697) [reply app/helpers/users_helper.rb]: done

**Pull request merged**